perf/ring.go

```diff
@@ -107,54 +117,186 @@ func createPerfEvent(cpu, watermark int) (int, error) {
 
 	n := len(p)
 	// Truncate if the read wraps in the ring buffer
-	if remainder := cap(rr.ring) - start; n > remainder {
+	if remainder := cap(fr.ring) - start; n > remainder {
 		n = remainder
 	}
 
 	// Truncate if there isn't enough data
-	if remainder := int(rr.head - rr.tail); n > remainder {
+	if remainder := int(fr.head - fr.tail); n > remainder {
 		n = remainder
 	}
 
-	copy(p, rr.ring[start:start+n])
-	rr.tail += uint64(n)
+	copy(p, fr.ring[start:start+n])
+	fr.tail += uint64(n)
 
-	if rr.tail == rr.head {
+	if fr.tail == fr.head {
 		return n, io.EOF
 	}
 
 	return n, nil
 }
+
+func (rr *reverseReader) loadHead() {
+	rr.previousHead = rr.head
+	rr.head = atomic.LoadUint64(&rr.meta.Data_head)
+	rr.read = atomic.LoadUint64(&rr.meta.Data_head)
+}
+
+func (rr *reverseReader) getRing() []byte {
+	return rr.ring
+}
+
+func (rr *reverseReader) writeTail() {
+	// We do not care about tail for over writable perf buffer.
+	// So, this function is noop.
+}
+
+func (rr *reverseReader) Read(p []byte) (int, error) {
+	read := rr.read
+
+	// This prevents reading data we already processed:
+	//
+	//    head        previous_head--+     read
+	//     |                         |      |
+	//     V                         V      V
+	// +---+------+----------+-------+------+
+	// |   |D....D|C........C|B.....B|A....A|
+	// +---+------+----------+-------+------+
+	// <--Write from right to left
+	//             Read from left to right-->
+	if rr.previousHead != 0 && read >= rr.previousHead {
+		rr.previousHead = rr.head
+
+		return 0, io.EOF
+	}
+
+	// If adding the size to the current consumer position makes us wrap the
+	// buffer, it means we already did "one loop" around the buffer.
+	// So, the pointed data would not be usable:
+	//
+	//                                  head
+	//                       read----+   |
+	//                               |   |
+	//                               V   V
+	// +---+------+----------+-------+---+--+
+	// |..E|D....D|C........C|B.....B|A..|E.|
+	// +---+------+----------+-------+---+--+
+	if read-rr.head+uint64(len(p)) > rr.mask {
+		rr.previousHead = rr.head
+
+		return 0, io.EOF
+	}
+
+	size := uint32(len(p))
+	previousReadMasked := read & rr.mask
+	read += uint64(size)
+
+	// If adding the event size to the current
+	// consumer position makes us going from end of the buffer toward the
+	// start, we need to copy the rr.ring in two times:
+	// 1. First from previous_read until end of the buffer.
+	// 2. Second from start of the buffer until read.
+	//
+	//    read                   previous_read
+	//     |                             |
+	//     V                             V
+	// +---+------+----------+-------+---+--+
+	// |..E|D....D|C........C|B.....B|A..|E.|
+	// +---+------+----------+-------+---+--+
+	// This code snippet was highly inspired by gobpf:
+	// https://github.com/iovisor/gobpf/blob/16120a1bf4d4abc1f9cf37fecfb86009a1631b9f/elf/perf.go#L148
+	if (read & rr.mask) < previousReadMasked {
```

Comment:
> Indeed, we want to start to read, from left to right, beginning from head. So, on the first call, this would be true and we will actually never read anything.

I think that is because you're initializing `head == read` in `readHead()`. See my larger comment.